### PR TITLE
Implement Adam Optimizer

### PR DESCRIPTION
**Pull request by haydnv**
@KOHbDS Here's a version of your Adam optimizer that works with the `optimizer.train` function. I removed the `dL` parameter from the `cost` function because I was concerned about optimizing a negative loss but PTAL and see if I made any mistakes in the loss to backpropagate (feel free to commit to this branch yourself).

### CODE REVIEW

File: client/tinychain/ml/dnn.py [hunk 1]
```diff
@@ -87,12 +102,15 @@ def backward(self, inputs, loss):
...
-                layer_gradients = []
+                param_list = []
                 for l in reversed(range(n)):
-                    loss, layer_gradient = self[l].backward(layer_inputs[l], loss)
+                    loss, layer_param_list = self[l].backward(layer_inputs[l], loss)
                     loss = loss.copy()
-                    layer_gradients.insert(0, layer_gradient)
+                    param_list.extend(layer_param_list)
+
+                return loss, param_list
```
[the remark this answers is not in the document]
Comment (haydnv) [reply client/tinychain/ml/dnn.py:111]: The reason I think DNN.backward needs to follow the same return signature as DNNLayer.backward is to cover the case of a complex model where the loss needs to be backpropagated further. For example, a computer vision model might have an RNN that depends on a DNN that depends on a CNN that takes the actual input.
